google/cloud/storage/internal/grpc_resumable_upload_session.cc

```diff
@@ -163,6 +86,67 @@ GrpcResumableUploadSession::last_response() const {
   return last_response_;
 }
 
+StatusOr<ResumableUploadResponse> GrpcResumableUploadSession::UploadGeneric(
+    std::string const& payload, bool final_chunk) {
+  CreateUploadWriter();
+
+  std::size_t const maximum_chunk_size =
+      google::storage::v1::ServiceConstants_Values_MAX_WRITE_CHUNK_BYTES;
+  std::string chunk;
+  auto flush_chunk = [&](bool has_more) {
+    if (chunk.size() < maximum_chunk_size && has_more) return true;
+    if (chunk.empty() && !final_chunk) return true;
+
+    google::storage::v1::InsertObjectRequest request;
+    request.set_upload_id(session_id_);
+    request.set_write_offset(next_expected_);
+    request.set_finish_write(false);
+
+    auto& data = *request.mutable_checksummed_data();
+    auto const n = chunk.size();
+    data.set_content(std::move(chunk));
+    chunk.clear();
+    data.mutable_crc32c()->set_value(crc32c::Crc32c(data.content()));
+
+    auto options = grpc::WriteOptions();
+    if (final_chunk && !has_more) {
+      // At this point we can set the full object checksums, there are two bugs
+      // for this:
+      // TODO(#4156) - compute the crc32c value inline
+      // TODO(#4157) - compute the MD5 hash value inline
+      request.set_finish_write(true);
+      options.set_last_message();
+    }
+
+    if (!upload_writer_->Write(request, options)) return false;
+    // After the first message, clear the object specification and checksums,
+    // there is no need to resend it.
+    request.clear_insert_object_spec();
+    request.clear_object_checksums();
+
+    next_expected_ += n;
+    return true;
+  };
+
+  std::size_t payload_offset = 0;
+  do {
+    chunk.reserve(maximum_chunk_size);
```

Comment:
Is there a reason the reserve is in this loop rather than up on line 96 right after `chunk` is declared?

Reply:
Yes. It is (potentially) moved-out in the `flush_chunk()` lambda, around line 107.

Reply:
I will move this around to try to make it clearer.

Reply:
Fixed.